### PR TITLE
[PM-10395] Add new item type ssh key

### PR DESCRIPTION
**Pull request by quexten**
## 🎟️ Tracking

Server: https://github.com/bitwarden/server/pull/4575
Add Item Type: https://github.com/bitwarden/clients/pull/10360
Add SSH Agent: https://github.com/bitwarden/clients/pull/10293
Add Import/Export: https://github.com/bitwarden/clients/pull/10529

Jira: https://bitwarden.atlassian.net/browse/PM-10395

## 📔 Objective

Add the new SSH key item type, consisting of private key, and derived fingerprint and public key. We are not generating these on the fly since we do not have the sdk available on all clients.

> [!NOTE]  
> This will be merged into a feature branch. This PR does not support key generation or import, thus testing will be done in the feature branch.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: apps/browser/src/_locales/en/messages.json [hunk 1]
```diff
@@ -1561,7 +1561,10 @@
   "typeIdentity": {
     "message": "Identity"
   },
-  "newItemHeader":{
+  "typeSSHKey": {
+    "message": "SSH Key"
```
Comment (chadmiller) [inline apps/browser/src/_locales/en/messages.json:1565]: Follow pattern of other strings, Lowercase words after the first. See "Secure note" and "Password history"

```suggestion
    "message": "SSH key"
```

File: apps/browser/src/vault/popup/components/vault/add-edit.component.html [hunk 2]
```diff
@@ -529,6 +529,20 @@ <h2 class="box-header">
...
         </div>
+
+        <!-- SSHKey -->
+        <div *ngIf="cipher.type === cipherType.SSHKey">
+          <div class="box-content-row" appBoxRow>
+            <label for="privateKey">{{ "privateKey" | i18n }}</label>
+            <input
+              id="privateKey"
+              type="text"
+              name="Card.CardCardholderName"
```
Comment (chadmiller) [inline apps/browser/src/vault/popup/components/vault/add-edit.component.html:540]: This is probably a nonpersonal identifier. Sally may have several such keys.

File: apps/browser/src/vault/popup/components/vault/view.component.html [hunk 3]
```diff
@@ -429,6 +429,27 @@ <h2 class="box-header">
...
         </div>
       </div>
+      <!-- SSHKey -->
+      <div *ngIf="cipher.sshKey">
+        <div class="box-content-row" *ngIf="cipher.sshKey.privateKey">
+          <span
+            class="row-label draggable"
+            draggable="true"
+            (dragstart)="setTextDataOnDrag($event, cipher.sshKey.privateKey)"
+            >Private Key</span
```
Comment (chadmiller) [inline apps/browser/src/vault/popup/components/vault/view.component.html:439]: i18n

File: apps/browser/src/vault/popup/components/vault/view.component.html [hunk 4]
```diff
@@ -429,6 +429,27 @@ <h2 class="box-header">
...
+          >
+          {{ cipher.sshKey.privateKey }}
+        </div>
+        <div class="box-content-row" *ngIf="cipher.identity.publicKey">
+          <span
+            class="row-label draggable"
+            draggable="true"
+            (dragstart)="setTextDataOnDrag($event, cipher.sshKey.publicKey)"
+            >Public Key</span
```
Comment (chadmiller) [inline apps/browser/src/vault/popup/components/vault/view.component.html:448]: i18n

File: apps/desktop/src/locales/en/messages.json [hunk 5]
```diff
@@ -174,6 +177,27 @@
...
+    "message": "SSH public key"
+  },
+  "sshFingerprint": {
+    "message": "SSH fingerprint"
+  },
+  "sshKeyAlgorithm": {
+    "message": "Key type"
+  },
+  "sshKeyAlgorithmED25519": {
+    "message": "ED25519 (Recommended)"
```
Comment (chadmiller) [inline apps/desktop/src/locales/en/messages.json:193]: "Recommended" is weird.

File: apps/desktop/src/locales/en/messages.json [hunk 6]
```diff
@@ -174,6 +177,27 @@
...
+  },
+  "sshKeyAlgorithmRSA2048": {
+    "message": "RSA 2048 bit"
+  },
+  "sshKeyAlgorithmRSA4096": {
+    "message": "RSA 4096 bit"
```
Comment (chadmiller) [inline apps/desktop/src/locales/en/messages.json:199]: Different sizes are not different types. Keys can be even more sizes than this, too.

File: libs/angular/src/vault/components/add-edit.component.ts [hunk 7]
```diff
@@ -178,6 +181,11 @@ export class AddEditComponent implements OnInit, OnDestroy {
       { name: i18nService.t("autoFillOnPageLoadYes"), value: true },
       { name: i18nService.t("autoFillOnPageLoadNo"), value: false },
     ];
+    this.sshKeyAlgorithmOptions = [
+      { name: i18nService.t("sshKeyAlgorithmED25519"), value: "ed25519" },
+      { name: i18nService.t("sshKeyAlgorithmRSA2048"), value: "rsa-2048" },
+      { name: i18nService.t("sshKeyAlgorithmRSA4096"), value: "rsa-4096" },
```
Comment (chadmiller) [inline libs/angular/src/vault/components/add-edit.component.ts:187]: - DSA
- ECDSA
- ED25519
- RSA

Comment (quexten) [reply libs/angular/src/vault/components/add-edit.component.ts:187]: I still need to discuss with the product team, which key algorithms and sizes will be supported. That being said I do not think DSA should be included.

https://linuxsecurity.com/news/security-projects/openssh-plans-to-remove-support-for-dsa-keys

File: apps/desktop/src/locales/en/messages.json (hunk 6 above)
Comment (quexten) [reply apps/desktop/src/locales/en/messages.json:199]: Same issue here. Not sure whether arbitrary bit sizes will be supported yet. I guess it is needed for import of key files, if the user has chosen uncommon key sizes.

File: libs/angular/src/vault/components/add-edit.component.ts (hunk 7 above)
Comment (quexten) [reply libs/angular/src/vault/components/add-edit.component.ts:187]: After digging a bit into this, ECDSA and especially DSA will not be supported. If a user has these keys they can generated ED25519 or RSA keys from within bitwarden and add them as extra keys on their servers / git signing.

File: apps/desktop/src/locales/en/messages.json (hunk 5 above)
Comment (quexten) [reply apps/desktop/src/locales/en/messages.json:193]: Will change this, to give info banners instead. Most likely something like:

Ed25519 is the most modern, and fast encryption method.
RSA is the most widely supported encryption method.

Ed25519 will be the default.

File: apps/desktop/src/vault/app/vault/view.component.html [hunk 8]
```diff
@@ -399,6 +399,55 @@ <h2 class="box-header">
             <div *ngIf="cipher.identity.country">{{ cipher.identity.country }}</div>
           </div>
         </div>
+        <!-- SSH Key -->
+        <div *ngIf="cipher.card">
```
Comment (gbubemismith) [inline apps/desktop/src/vault/app/vault/view.component.html:403]: This should be cipher.sshKey

File: libs/angular/src/vault/components/add-edit.component.ts [hunk 9]
```diff
@@ -128,6 +129,11 @@ export class AddEditComponent implements OnInit, OnDestroy {
       { name: i18nService.t("typeIdentity"), value: CipherType.Identity },
       { name: i18nService.t("typeSecureNote"), value: CipherType.SecureNote },
     ];
+
+    if (this.platformUtilsService.getClientType() == ClientType.Desktop) {
```
[the remark this answers is not in the document]
Comment (quexten) [reply libs/angular/src/vault/components/add-edit.component.ts:133]: Hmm, not sure why it made it to this PR, it should only be in the followup PR (including generator / import), where it *is* featureflagged. I will remove it here

File: libs/common/src/vault/models/domain/ssh-key.ts
[the remark this answers is not in the document]
Comment (quexten) [reply libs/common/src/vault/models/domain/ssh-key.ts:1]: Added tests

**Review by gbubemismith — APPROVED** [review]

**Comment by quexten** [comment]
Merging to the feature branch; QA testing will be done on the feature-branch.

**Pull request merged**